metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala

```diff
@@ -668,6 +676,34 @@ object MetalsEnrichments
     def asTextEdit: Option[l.TextEdit] = {
       decodeJson(d.getData, classOf[l.TextEdit])
     }
+    object EitherDeserializer
+        extends JsonDeserializer[Either[l.TextEdit, DiagnosticData]] {
+      override def deserialize(
+          json: JsonElement,
+          typeOfT: Type,
+          context: JsonDeserializationContext,
+      ): Either[l.TextEdit, DiagnosticData] = {
+        json match {
+          case o: JsonObject if o.has("edits") =>
+            Right(new Gson().fromJson(o, classOf[DiagnosticData]))
+          case o => Left(new Gson().fromJson(o, classOf[l.TextEdit]))
+        }
+      }
+    }
+
+    def asDiagnosticData: Option[Either[l.TextEdit, DiagnosticData]] = {
+      val gson = new GsonBuilder()
+        .registerTypeAdapter(
+          classOf[Either[l.TextEdit, DiagnosticData]],
+          EitherDeserializer,
+        )
+        .create()
+      decodeJson(
+        d.getData,
+        classOf[Either[l.TextEdit, DiagnosticData]],
+        Some(gson),
+      )
+    }
```

Comment:
So this is a bit weird, but the way gson parses here tripped me up a bit. I think that we could first try to parse it as a `l.TextEdit` and if that fails, then try as a `DiagnosticData`, but it turns out that because `DiagnosticData` _contains_ a `WorkspaceEdit` that has some of the same fields as `TextEdit` it ends up parsing every time as a `TextEdit` with `null` values. To illustrate what I mean

```scala
import $dep.`com.google.code.gson:gson:2.10.1`
import com.google.gson.Gson
import com.google.gson.JsonElement

case class Foo(a: Int, b: Int)
case class Bar(c: Int, foo: Foo)

val foo = Foo(1, 2)
val bar = Bar(1, foo)

val json = new Gson().toJsonTree(foo)
val otherJson = new Gson().toJsonTree(bar)

new Gson()
  .fromJson[Foo](otherJson.asInstanceOf[JsonElement], classOf[Foo])
```
Since you're trying to parse this into `Foo` but giving it a `Bar` instead of failing to parse it will just give you `Foo(0, 0)`. Because of this I ended up going with a custom either parser to check if `edits` exist in the json element. If it does, then we parse it as a `DiagnosticData` or else we fall back to a `TextEdit`.

Reply:
Could we potentially have `DiagnosticData` as Java Class with a `@NotNull` annotation? Wouldn't that make gson happy?